### PR TITLE
Only load default config when saving test data from snmpsim

### PR DESCRIPTION
**Pull request by TheMysteriousX**
If you're generating test data from a real device, you're probably
doing so on a production instance (or at least, that's where I
poll all of my most expensive kit from!).

This means there are likely to be a bunch of config settings that
cause test data to be generated differently between CI and your
own instance.

This change clears the if, mount and sensor filters before test
data is retrieved from SNMPSim.

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: LibreNMS/Config.php [hunk 1]
```diff
@@ -133,6 +133,24 @@ public static function get($key, $default = null)
...
+    /**
+     * Find the default value for a configuration item
+     *
+     * @param string $key period separated config variable name
+     * @param mixed $default optional value to return if the setting is not set
+     * @return mixed
+     */
+    public static function getDefault($key, $default = null)
+    {
+        $definitions = self::getDefinitions();
```
Comment (murrant) [inline LibreNMS/Config.php:145]: getDefinitions() is an expensive operation.  Doing it like this is going to parse json for every call.

**Comment by murrant** [comment]
Been meaning to do something like this for awhile.

I was more thinking to load the settings without config.php would be my approach.

**Comment by TheMysteriousX** [comment]
I think you're right, changing the initial config loading is probably a better approach than trying to fix it up once it's happened.

**Comment by TheMysteriousX** [comment]
I think this should do the trick, given it a quick test on my live instance that was generating broken test data, and it seems to be working.

Not sure about the name "testMode" - if you have any better suggestions I can change it.

**Comment by murrant** [comment]
Not bad.

What do you think about having a function loadDefaults().  That way we don't need to have a check in the normal load process.

**Comment by TheMysteriousX** [comment]
Will test it in a bit, but the latest change removes the condition.

**Comment by Jellyfrog** [comment]
But don't we want \*all\* modules to be enabled for snmpsim? (Especially for `.snmprec`)

**Comment by TheMysteriousX** [comment]
@Jellyfrog That isn't done right now.

Without this change, the test .json files are saved with user config customisations - which also included disabled modules.

With this change, user customisations are ignored when saving the test data - i.e. it makes sure that the environment used to collect data and Travis match.

If all modules should be enabled for this, that could be done, but my feeling is that it should be.
The modules that are disabled in the default configuration are usually done so for a good reason (i.e. your iDRAC is never going to start talking BGP or sprout a wireless access point). But even in that case, you'd still want user customisations ignored when doing the collection.

@murrant I had to revise things again - at the point loadDefault() was called, Laravel isn't bootstrapped enough to actually load the config. In this iteration, I've used a PHP's equivalent of a function pointer to specify which set of config to load.

**Comment by murrant** [comment]
Hmm, not working when I set `$config['bad_if_regexp'][] = '/.*/';` in my config.php

**Comment by murrant** [comment]
How about just this:
```diff
--- LibreNMS/Config.php	(revision c31d58f9bbd636cf2dcbb1a55761823034a82546)
+++ LibreNMS/Config.php	(date 1587821925640)
@@ -51,8 +51,11 @@
 
         // merge all config sources together config_definitions.json > db config > config.php
         self::loadDefaults();
-        self::loadDB();
-        self::loadUserConfigFile(self::$config);
+
+        if (app()->environment() !== 'testing') {
+            self::loadDB();
+            self::loadUserConfigFile(self::$config);
+        }
 
         // final cleanups and validations
         self::processConfig();
```

**Comment by TheMysteriousX** [comment]
I'll give it a try. Not sure why you're seeing that bad_if_regexp was still set, maybe it's being leaked through the legacy $config global somehow?

**Comment by TheMysteriousX** [comment]
Setting app->env to testing doesn't appear to break anything (even though I've not got a .env file for 'testing', so I'll update this changeset with that, and add a prompt to the collection and pre-commit scripts noting that for correct behaviour, the app environment may need to be set to testing.

Re: bad_if_regexp, the leak is coming from the unit tests. Config::load gets called a lot.

File: scripts/pre-commit.php [hunk 2]
```diff
@@ -336,6 +336,10 @@ function check_style($passthru = false, $command_only = false)
  */
 function check_unit($passthru = false, $command_only = false, $options = array())
 {
+    if (getenv('APP_ENV') !== 'testing') {
```
Comment (murrant) [inline scripts/pre-commit.php:339]: Or you could just `putenv('APP_ENV=testing')`

**Comment by murrant** [comment]
Ok, so some tests require the database to be loaded across tests... (which is probably an error in the testing)

Without fixing the tests, this probably brings us back to something you had before.

**Comment by TheMysteriousX** [comment]
I'll have a look and see how many tests are affected.

**Comment by murrant** [comment]
It is just this test, because the webserver and the tests run in different processes.

**Comment by murrant** [comment]
Has not solve issue where webserver tests run in a different process, so they need the db config loaded.  Feel free to re-open with a fix.

**Pull request closed without merging**